text/0000-futures.md

```diff
@@ -154,31 +162,315 @@ can be dropped.
+specialized executors, which e.g. want to avoid the cost of all synchronization.
+
+Calling the `wake()` method on a `Waker` will in general be more expensive than
+on a `LocalWaker` instance, due to additional synchronization.
+
+Executors will always pass a `LocalWaker` instance to the tasks they poll.
+
+The mechanism through which tasks get scheduled again depends on the executor
+which is driving the task.
+Possible ways of waking up a an executor include:
+- If the executor is blocked on a condition variable, the condition variable
+  needs to get notified.
+- If the executor is blocked on a system call like `select`, it might need
+  to get woken up by a syscall like `write` to a pipe.
+- If the executor's thread is parked, the wakeup call needs to unpark it.
+
+In order to accomodate for this behavior, the `Waker` and `LocalWaker` types are
+defined through a `RawWaker` type, which is an struct that defines a dynamic
+dispatch mechanism, which consists of an raw object pointer and a virtual function
+pointer table (vtable). This mechanism allows implementors of an executor to
+customize the behavior of `RawWaker`, `Waker` and `LocalWaker` objects.
+
+This mechanism is chosen in favor of trait objects since it allows for more
+flexible memory management schemes. `RawWaker` can be implemented purely in
+terms of global functions and state, on top of reference counted objects, or
+in other ways.
+
+The relation between those `Waker` types is outlined in the following definitions:
+
+```rust
+/// A `RawWaker` allows the implementor of a task executor to customize the
+/// behavior of `LocalWaker`s and `Waker`s.
+/// It consists of a data pointer a virtual function pointer table (vtable) that
+/// customizes the behavior of the `RawWaker`.
+#[derive(PartialEq)]
+pub struct RawWaker {
+    /// A data pointer, which can be used to store arbitrary data as required
+    /// by the executor. This could be e.g. a type-erased pointer to an `Arc`
+    /// that is associated with the task.
+    /// The value of this field gets passed to all functions that are part of
+    /// the vtable as first parameter.
+    pub data: *const (),
+    /// Virtual function pointer table that customizes the behavior if this waker.
+    pub vtable: &'static RawWakerVTable,
+}
+
+/// A virtual function pointer table (vtable) that allows to customize the
+/// behavior of a `RawWaker`.
+/// The pointers which get passed to all functions inside the vtable are the
+/// values of the `data` field.
+#[derive(PartialEq, Copy, Clone)]
+pub struct RawWakerVTable {
+    /// This function will be called when the `RawWaker` gets cloned, e.g. when
+    /// the `Waker` or `LocalWaker` in which the `RawWaker` is stored gets cloned.
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a `RawWaker` and associated
+    /// task.
+    pub clone: unsafe fn(*const ()) -> RawWaker,
```

Comment:
I really wish we had `unsafe` closures, then this could be implemented something like

```rust
static VTABLE: RawWakerVTable = RawWakerVTable {
    clone: unsafe |data| (data as *const ActualWaker as &'static ActualWaker).clone(),
    ...
}
```

Reply:
I had the same realization while implementing things. But it's more a cosmetic issue for very trivial use-cases than a blocker.

Reply:
Yep, this was more a comment in support of `unsafe` closures in case they're ever proposed and I remember to link here than a downside of this change. I managed to workaround it with

```rust
clone: {
        |data| (data as *const EmbrioWaker as &'static EmbrioWaker).raw_waker()
} as fn(*const ()) -> RawWaker as unsafe fn(*const ()) -> RawWaker,
```